sql/src/main/java/org/apache/druid/sql/calcite/run/SqlResults.java

```diff
@@ -0,0 +1,309 @@
+          }
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+        }
+      } else {
+        // the protobuf jdbc handler prefers lists (it actually can't handle java arrays as sql arrays, only java lists)
+        // the json handler could handle this just fine, but it handles lists as sql arrays as well so just convert
+        // here if needed
+        coercedValue = maybeCoerceArrayToList(value, true);
+        if (coercedValue == null) {
+          throw new ISE("Cannot coerce [%s] to %s", value.getClass().getName(), sqlTypeName);
+        }
+      }
+    } else {
+      throw new ISE("Cannot coerce [%s] to %s", value.getClass().getName(), sqlTypeName);
+    }
+
+    return coercedValue;
+  }
+
+
+  @VisibleForTesting
+  static Object maybeCoerceArrayToList(Object value, boolean mustCoerce)
+  {
+    if (value instanceof List) {
+      return value;
+    } else if (value instanceof String[]) {
+      return Arrays.asList((String[]) value);
+    } else if (value instanceof Long[]) {
+      return Arrays.asList((Long[]) value);
+    } else if (value instanceof Double[]) {
+      return Arrays.asList((Double[]) value);
+    } else if (value instanceof Object[]) {
+      final Object[] array = (Object[]) value;
+      final ArrayList<Object> lst = new ArrayList<>(array.length);
+      for (Object o : array) {
+        lst.add(maybeCoerceArrayToList(o, false));
+      }
+      return lst;
+    } else if (value instanceof long[]) {
+      return Arrays.stream((long[]) value).boxed().collect(Collectors.toList());
+    } else if (value instanceof double[]) {
+      return Arrays.stream((double[]) value).boxed().collect(Collectors.toList());
+    } else if (value instanceof float[]) {
+      final float[] array = (float[]) value;
+      final ArrayList<Object> lst = new ArrayList<>(array.length);
+      for (float f : array) {
+        lst.add(f);
+      }
+      return lst;
+    } else if (value instanceof ComparableStringArray) {
+      return Arrays.asList(((ComparableStringArray) value).getDelegate());
+    } else if (value instanceof ComparableList) {
+      return ((ComparableList) value).getDelegate();
+    } else if (mustCoerce) {
+      return null;
+    }
+    return value;
+  }
+
+  private static DateTime coerceDateTime(Object value, SqlTypeName sqlType)
+  {
+    final DateTime dateTime;
+
+    if (value instanceof Number) {
+      dateTime = DateTimes.utc(((Number) value).longValue());
+    } else if (value instanceof String) {
+      dateTime = DateTimes.utc(Long.parseLong((String) value));
+    } else if (value instanceof DateTime) {
+      dateTime = (DateTime) value;
+    } else {
+      throw new ISE("Cannot coerce[%s] to %s", value.getClass().getName(), sqlType);
+    }
+    return dateTime;
+  }
+
+  /**
+   * Context for {@link #coerce(ObjectMapper, Context, Object, SqlTypeName)}
+   */
+  public static class Context
```

Comment:
nit: Since we have so many contexts like queryContext the name `Context` might get confusing. Should we rename this to SqlResultParams so that the callers are not confused.

Reply:
IMO, this one is OK, since it's an inner static class of `SqlResults`. So call sites would generally refer to it as `SqlResults.Context`. (We usually don't do `import static` for inner classes.)